In [93]:
#Import neccesary libraries
import pandas as pd 
import numpy as np 

In [94]:
#Import data using pandas 
df= pd.read_csv('/kaggle/input/used-cars-prices-in-uk/used_cars_UK.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3685 entries, 0 to 3684
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         3685 non-null   int64  
 1   title              3685 non-null   object 
 2   Price              3685 non-null   int64  
 3   Mileage(miles)     3685 non-null   int64  
 4   Registration_Year  3685 non-null   int64  
 5   Previous Owners    2276 non-null   float64
 6   Fuel type          3685 non-null   object 
 7   Body type          3685 non-null   object 
 8   Engine             3640 non-null   object 
 9   Gearbox            3685 non-null   object 
 10  Doors              3660 non-null   float64
 11  Seats              3650 non-null   float64
 12  Emission Class     3598 non-null   object 
 13  Service history    540 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 403.2+ KB


In [95]:
#Exploring data 
df.head(5)

Unnamed: 0           title  Price  Mileage(miles)  Registration_Year  \
0           0     SKODA Fabia   6900           70189               2016   
1           1  Vauxhall Corsa   1495           88585               2008   
2           2     Hyundai i30    949          137000               2011   
3           3      MINI Hatch   2395           96731               2010   
4           4  Vauxhall Corsa   1000           85000               2013   

   Previous Owners Fuel type  Body type Engine Gearbox  Doors  Seats  \
0              3.0    Diesel  Hatchback   1.4L  Manual    5.0    5.0   
1              4.0    Petrol  Hatchback   1.2L  Manual    3.0    5.0   
2              NaN    Petrol  Hatchback   1.4L  Manual    5.0    5.0   
3              5.0    Petrol  Hatchback   1.4L  Manual    3.0    4.0   
4              NaN    Diesel  Hatchback   1.3L  Manual    5.0    5.0   

  Emission Class Service history  
0         Euro 6             NaN  
1         Euro 4            Full  
2         Euro 5             NaN  
3         Euro 4            Full  
4         Euro 5             NaN

In [96]:
df.describe()

Unnamed: 0         Price  Mileage(miles)  Registration_Year  \
count  3685.000000   3685.000000    3.685000e+03        3685.000000   
mean   2314.770963   5787.145726    8.132816e+04        2011.835007   
std    1415.821308   4480.810572    3.942083e+04           5.092566   
min       0.000000    400.000000    1.000000e+00        1953.000000   
25%    1059.000000   2490.000000    5.698400e+04        2008.000000   
50%    2279.000000   4000.000000    8.000000e+04        2012.000000   
75%    3593.000000   7995.000000    1.030000e+05        2015.000000   
max    4727.000000  33900.000000    1.110100e+06        2023.000000   

       Previous Owners        Doors        Seats  
count      2276.000000  3660.000000  3650.000000  
mean          2.807557     4.321038     4.900274  
std           1.546028     0.986902     0.577200  
min           1.000000     2.000000     2.000000  
25%           2.000000     3.000000     5.000000  
50%           3.000000     5.000000     5.000000  
75%           4.000000     5.000000     5.000000  
max           9.000000     5.000000     7.000000

In [97]:
#Checking data validation
null_values = df.isnull().sum()
null_values

Unnamed: 0              0
title                   0
Price                   0
Mileage(miles)          0
Registration_Year       0
Previous Owners      1409
Fuel type               0
Body type               0
Engine                 45
Gearbox                 0
Doors                  25
Seats                  35
Emission Class         87
Service history      3145
dtype: int64

In [98]:
#filling null value for doors with the mean 
door_means = df['Doors'].mean()
door_mean= door_means.round().astype(int)
doors = df['Doors'].fillna(door_mean,inplace=True)
df.isnull().sum()

Unnamed: 0              0
title                   0
Price                   0
Mileage(miles)          0
Registration_Year       0
Previous Owners      1409
Fuel type               0
Body type               0
Engine                 45
Gearbox                 0
Doors                   0
Seats                  35
Emission Class         87
Service history      3145
dtype: int64

In [99]:
#filling null value for seats with the mean 
seat_means = df['Seats'].mean()
seat_mean= seat_means.round().astype(int)
seat = df['Seats'].fillna(seat_mean,inplace=True)
df.isnull().sum()

Unnamed: 0              0
title                   0
Price                   0
Mileage(miles)          0
Registration_Year       0
Previous Owners      1409
Fuel type               0
Body type               0
Engine                 45
Gearbox                 0
Doors                   0
Seats                   0
Emission Class         87
Service history      3145
dtype: int64

In [100]:
duplicated_rows = df.duplicated().sum()
duplicated_rows

0

In [101]:
#importing sklearn models 
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

In [102]:
lr=linear_model.LinearRegression()
lr

LinearRegression()

In [103]:
df.columns

Index(['Unnamed: 0', 'title', 'Price', 'Mileage(miles)', 'Registration_Year',
       'Previous Owners', 'Fuel type', 'Body type', 'Engine', 'Gearbox',
       'Doors', 'Seats', 'Emission Class', 'Service history'],
      dtype='object')

In [104]:
#Checking correlations 
variables = df[['Price','Mileage(miles)','Registration_Year','Doors', 'Seats',]]
variables.corr()['Price']

Price                1.000000
Mileage(miles)      -0.500158
Registration_Year    0.723880
Doors                0.113061
Seats               -0.027513
Name: Price, dtype: float64

In [108]:
df_2 = df[['Price','Mileage(miles)','Registration_Year','Doors','Seats']]
df_2.head(10)

Price  Mileage(miles)  Registration_Year  Doors  Seats
0   6900           70189               2016    5.0    5.0
1   1495           88585               2008    3.0    5.0
2    949          137000               2011    5.0    5.0
3   2395           96731               2010    3.0    4.0
4   1000           85000               2013    5.0    5.0
5    800          124196               2007    3.0    4.0
6    798          140599               2008    5.0    5.0
7   1995           90000               2009    3.0    5.0
8    750          225318               1989    5.0    5.0
9   1299           87000               2008    5.0    5.0

In [109]:
independent_variables = df[['Mileage(miles)','Registration_Year','Doors','Seats']].to_numpy()
independent_variables

array([[7.0189e+04, 2.0160e+03, 5.0000e+00, 5.0000e+00],
       [8.8585e+04, 2.0080e+03, 3.0000e+00, 5.0000e+00],
       [1.3700e+05, 2.0110e+03, 5.0000e+00, 5.0000e+00],
       ...,
       [1.3900e+05, 2.0130e+03, 4.0000e+00, 5.0000e+00],
       [1.7919e+05, 2.0070e+03, 5.0000e+00, 5.0000e+00],
       [8.2160e+04, 2.0130e+03, 5.0000e+00, 5.0000e+00]])

In [110]:
target_variable = df['Price'].to_numpy()
target_variable

array([6900, 1495,  949, ..., 3995, 1390, 2000])

Training The Model 

In [111]:
x=independent_variables 
y=target_variable

In [115]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=500, random_state=5)

In [116]:
y_train

array([3195, 6500, 6500, ..., 1595, 1995, 9450])

In [117]:
lr.fit(x_train,y_train)

LinearRegression()

In [118]:
correlaation_coefficient = lr.coef_
correlaation_coefficient

array([-2.63622765e-02,  5.43229411e+02, -7.31375728e+01, -1.20242442e+02])

In [121]:
y_prediction = lr.predict(x_test)
y_prediction
df_3 = pd.DataFrame(y_prediction, columns=['Prediction'])
df_3.head(100)

Prediction
0    8434.357129
1    5688.617169
2    8422.994987
3    8009.748282
4    4196.689576
5    5354.567494
6    5066.300929
7    8812.893057
8    2044.184952
9    5734.017761
10  12544.725900
11   7421.888090
12   2136.752585
13   -854.997639
14   8306.798690
15   8967.031949
16   6404.611947
17   4686.850662
18   6308.038587
19   3597.840790
20   4950.399645
21   9570.737761
22   5274.401150
23   5437.819564
24   1108.232057
25   5250.666904
26   8641.009676
27   6271.641686
28   8256.675385
29  11597.336428
30   7797.905680
31   3721.654867
32   7757.348844
33   3445.272616
34   4699.818651
35   1517.013858
36  10562.951444
37   6176.570976
38   7445.227832
39  11437.501945
40  -3612.907892
41    919.268598
42   3291.528096
43  10412.238309
44   3058.696746
45   4098.724043
46  -6539.325851
47   9646.845653
48   8171.147819
49   6445.799226
50   5437.819564
51  11298.933479
52   5533.497945
53   5740.942698
54   4512.919790
55  11339.636834
56  10152.877891
57  11235.118749
58   5988.940975
59   4271.110559
60   9362.466097
61   7447.565733
62   3106.992161
63   7254.602487
64   3866.569495
65   1975.550954
66   6492.162129
67   2849.152137
68   6508.558403
69   5553.572630
70   7361.074417
71   3433.145969
72  10412.238309
73   6123.846423
74   3893.431317
75   1919.842945
76   7469.155099
77   5525.500639
78   8281.403200
79   5857.604113
80   6133.881047
81   8134.863647
82   8812.893057
83   3731.365720
84   8815.265662
85   3278.857244
86  11532.203584
87   9704.494923
88   4933.452800
89   7500.290286
90   8281.403200
91  11250.356145
92   3411.784737
93   2242.961545
94   4209.778634
95   8998.798492
96   7445.227832
97   3616.719493
98   8629.517061
99  10571.044663

In [122]:
df_2['Price Prediction']=df_3

/tmp/ipykernel_32/1279879697.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [123]:
pd.set_option('display.max_rows',None)
df_2.head(100)

Price  Mileage(miles)  Registration_Year  Doors  Seats  Price Prediction
0    6900           70189               2016    5.0    5.0       8434.357129
1    1495           88585               2008    3.0    5.0       5688.617169
2     949          137000               2011    5.0    5.0       8422.994987
3    2395           96731               2010    3.0    4.0       8009.748282
4    1000           85000               2013    5.0    5.0       4196.689576
5     800          124196               2007    3.0    4.0       5354.567494
6     798          140599               2008    5.0    5.0       5066.300929
7    1995           90000               2009    3.0    5.0       8812.893057
8     750          225318               1989    5.0    5.0       2044.184952
9    1299           87000               2008    5.0    5.0       5734.017761
10   1495          181693               2009    3.0    5.0      12544.725900
11   2950           79000               2011    5.0    5.0       7421.888090
12   6900           70189               2016    5.0    5.0       2136.752585
13  19800           69690               2019    4.0    5.0       -854.997639
14   1100          164985               2008    3.0    4.0       8306.798690
15   1295          109000               2007    5.0    5.0       8967.031949
16    999          111400               2007    5.0    5.0       6404.611947
17   1289          135000               2010    5.0    5.0       4686.850662
18   1200          134000               2004    5.0    5.0       6308.038587
19   1949          114000               2013    5.0    5.0       3597.840790
20   1195          101000               2010    5.0    5.0       4950.399645
21  14498           53055               2019    5.0    5.0       9570.737761
22   1995           89000               2007    2.0    4.0       5274.401150
23    990           69000               2003    3.0    4.0       5437.819564
24   2950           79000               2011    5.0    5.0       1108.232057
25  15900           62402               2019    5.0    5.0       5250.666904
26   8975           86500               2017    5.0    5.0       8641.009676
27   3399           55000               2011    5.0    5.0       6271.641686
28   1990          150000               2007    5.0    5.0       8256.675385
29   1200           95000               2009    5.0    5.0      11597.336428
30   1490          135800               2008    5.0    5.0       7797.905680
31   2250           98000               2006    2.0    4.0       3721.654867
32   2800          106305               2011    5.0    5.0       7757.348844
33   2490           86000               2011    2.0    4.0       3445.272616
34   1695           75786               2008    3.0    5.0       4699.818651
35   1795           99905               2010    5.0    5.0       1517.013858
36   2500           95000               2011    5.0    5.0      10562.951444
37   2495          149000               2010    5.0    5.0       6176.570976
38  15900           62402               2019    5.0    5.0       7445.227832
39  10698           22782               2020    5.0    5.0      11437.501945
40   1795           63796               2008    3.0    5.0      -3612.907892
41   1250          178000               2007    5.0    5.0        919.268598
42    988          113218               2005    3.0    5.0       3291.528096
43   1995           88000               2008    5.0    5.0      10412.238309
44   1999          122000               2009    5.0    5.0       3058.696746
45   1449           57000               1999    4.0    5.0       4098.724043
46   1290          119000               2002    5.0    5.0      -6539.325851
47   1449          111400               2008    5.0    5.0       9646.845653
48   3295           94000               2014    3.0    5.0       8171.147819
49   1400          155085               2004    5.0    5.0       6445.799226
50   1950           93000               2012    3.0    5.0       5437.819564


In [124]:
coefficient_determination = r2_score(y_test,y_prediction)
coefficient_determination

0.6029476028883652

# Conclusion 
The model uses factors like mileage, registration year, the number of doors, and the number of seats to make its predictions.

The R-squared value of around 0.6029 means that our model can explain about 60.29% of the variability in car prices, which is not bad at all. However, there are still some factors or variations that our model did not account for.

Please kindly upvote if you this looks good to you 